pkg/assetfetcher/http.go

```diff
@@ -0,0 +1,155 @@
+package assetfetcher
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"sync"
+
+	"github.com/asecurityteam/nexpose-vuln-notifier/pkg/domain/nexpose"
+)
+
+// This const block helps format our request to the Nexpose sites asset endpoint:
+// GET /api/3/sites/{id}/assets (doc: https://help.rapid7.com/insightvm/en-us/api/index.html#operation/getSiteAssets)
+const (
+	basePath       = "/api/3"
+	nexposeSite    = "/sites"
+	nexposeAssets  = "/assets"
+	pageQueryParam = "page" // The index of the page (zero-based) to retrieve.
+	sizeQueryParam = "size" // The number of records per page to retrieve.
+
+)
+
+// SiteAssetsResponse is the structure of the Nexpose site assets response
+type SiteAssetsResponse struct {
+	// Hypermedia links to corresponding or related resources
+	Links nexpose.Link
+	// The details of pagination indicating which page was returned, and how the remaining pages can be retrieved.
+	Page Page
+	// The page of resources returned (resources = assets)
+	Resources []nexpose.Asset
+}
+
+// Page represents the JSON object Nexpose provides to help paginate through all the Assets
+type Page struct {
+	// The index (zero-based) of the current page returned.
+	Number int
+	// The maximum size of the page returned.
+	Size int
+	// The total number of pages available.
+	TotalPages int
+	// The total number of resources available across all pages.
+	TotalResources int
+}
+
+// NexposeAssetFetcher is used to create a new
+type NexposeAssetFetcher struct {
+	// The Nexpose HTTP Cient
+	HTTPClient *http.Client
+	// The Nexpose host that points to your instance
+	Host string
+	// The number of assets that should be returned at one time
+	PageSize int
+}
+
+// FetchAssets gets all the assets for a given site ID from Nexpose, with pagination.
+// It returns a channel of assets and a channel of errors that can by asynchronously listened to.
+func (c *NexposeAssetFetcher) FetchAssets(ctx context.Context, siteID string) (<-chan nexpose.Asset, <-chan error) {
+	var errChan = make(chan error)
```

Comment:
I think you will want to use buffered channels here, otherwise your writes to the channel will be blocked until something reads from the channel. https://stackoverflow.com/questions/18660533/why-using-unbuffered-channel-in-the-same-goroutine-gives-a-deadlock

Reply:
This code is structured so that a buffer is not needed. FetchAssets will make the channels and return them immediately, then the next lines in nexposevulnnotifier.Handle will start reading from the channel until the channels close.

Reply:
The problem with this is that you require the consumer of the method to _always read all values from all channels_. Even if one of the requests immediately returns an error, this method places the burden on the consumer to wait for all other requests to complete, successfully or otherwise, before it can continue.

Even a perfectly behaved consumer of this method may fault if it encountered a `panic`. In this case, the consumer would exit abruptly without consuming everything from the channels. As a result, you cannot guarantee in this method that every goroutine started will exit.

Since you know both the number of possible errors as well as the maximum number of possible assets, why not buffer the channels and let the goroutines exit as soon as they are done with their work?

Reply:
> Even if one of the requests immediately returns an error, this method places the burden on the consumer to wait for all other requests to complete, successfully or otherwise, before it can continue.

The intention here and my understanding of the code is that if an error is returned, then the error will be placed on the errChan and the consumer will have access to the error immediately and it won't have to wait for all the other requests to complete in order for it to continue. 

Either way, you're right, we do know the total number of assets we'll need, so I can add the buffer.